modules/gapi/test/gapi_statefulkernel_tests.cpp

```diff
@@ -0,0 +1,201 @@
+//     cv::Mat sourceImage(imageSize, CV_8UC1),
+//             differentImage(imageSize, CV_8UC1);
+//     cv::randu(sourceImage, cv::Scalar(0), cv::Scalar(255));
+//     cv::randu(differentImage, cv::Scalar(0), cv::Scalar(255));
+
+//     std::vector<cv::Mat> stream(3, sourceImage);
+
+//     bool isChanged = true;
+
+//     GMat in;
+//     auto out = GHasSourceChanged::on(in);
+//     cv::GComputation comp(cv::GIn(in), cv::GOut(out));
+
+//     const auto pkg = cv::gapi::kernels<GOCVHasSourceChanged>();
+//     comp.apply(cv::gin(stream[0]), cv::gout(isChanged), cv::compile_args(pkg));
+//     EXPECT_FALSE(isChanged);
+
+//     for (const auto &frame: stream)
+//     {
+//         comp.apply(cv::gin(frame), cv::gout(isChanged));
+//         EXPECT_FALSE(isChanged);
+//     }
+
+//     comp.apply(cv::gin(differentImage), cv::gout(isChanged));
+//     EXPECT_TRUE(isChanged);  
+//}
+
+TEST(StatefuKernel, StateIsMutableInRuntime)
+{ 
+    constexpr int expectedCallsCount = 10;
+
+    //graph with empty mat shouldn't be compiled - TO CREATE A TICKET 
+    cv::Mat dummyIn;
+    // to init in other tests
+    int actualCallsCount = 0;
+
+    // Declaration of G-API expression
+    GMat in;
+    GOpaque<int> out = GCountCalls::on(in);
+    cv::GComputation comp(cv::GIn(in), cv::GOut(out));
+
+    const auto pkg = cv::gapi::kernels<GOCVCountCalls>();
+
+    // Compilation of G-API expression
+    auto callsCounter = comp.compile(cv::descr_of(dummyIn), cv::compile_args(pkg));
+
+    // Streaming
+    for (int i = 0; i < expectedCallsCount; i++)
+    {
+        callsCounter(cv::gin(dummyIn), cv::gout(actualCallsCount));
+        EXPECT_EQ(i + 1, actualCallsCount);
+    }
+    // End of "stream"
+    EXPECT_EQ(expectedCallsCount, actualCallsCount);
+
+    // User tells G-API to reset for a new stream 
+    callsCounter.prepareForNewStream();
+    callsCounter(cv::gin(dummyIn), cv::gout(actualCallsCount));
+    EXPECT_EQ(1, actualCallsCount);
+      
+}
+//-------------------------------------------------------------------------------------------------------------
+
+                                       // std::vector doesn't work with GOpaque because of explicit 
+                                       // cv::descr_of instantiation in GArray.
+                                       //std::tuple<std::vector<int>, StdVectorValue>,
+
+//----------------------------------------- Typed tests on iniState() -----------------------------------------
+namespace
+{
+template<typename Tuple>
+struct InitStateTypedTest : public ::testing::Test
+{
+    using StateT = typename std::tuple_element<0, Tuple>::type;
+    using InitT  = typename std::tuple_element<1, Tuple>::type;
+
+    G_TYPED_KERNEL(GReturnState, <cv::GOpaque<StateT>(GMat)>, "org.opencv.test.return_state")
+    {
+        static GOpaqueDesc outMeta(GMatDesc /* in */) { return empty_gopaque_desc(); }
+    };
+
+    GAPI_OCV_STATEFUL_KERNEL(GOCVReturnState, GReturnState, StateT)
```

Comment:
So our kernels can be templates? this is cool :)

Reply:
Yes, I also was pleasantly surprised:)